src/code/m_npc_walk.c

```diff
@@ -0,0 +1,683 @@
+            break;
+
+        case mNpcW_GOAL_ALONE:
+            mNpcW_GetAloneBlock(&info->goal_block_x, &info->goal_block_z);
+            break;
+
+        case mNpcW_GOAL_MY_HOME:
+            idx = info->idx;
+
+            info->goal_block_x = common_data.save.animals[idx].home_info.block_x;
+            info->goal_block_z = common_data.save.animals[idx].home_info.block_z;
+            break;
+
+        default:
+            info->goal_block_x = 0;
+            info->goal_block_z = 0;
+            break;
+    }
+}
+
+void mNpcW_InitNpcWalk(mNpc_walk_c* walk) {
+    Animal_c* animals;
+    mNpcW_info_c* info = walk->info;
+    s32 idx;
+    s32 i;
+
+    mNpcW_ClearNpcWalk(walk);
+    bzero(l_goal_block, sizeof(l_goal_block));
+    bzero(l_arrive_stay_count, sizeof(l_arrive_stay_count));
+
+    animals = common_data.save.animals;
+    walk->info_max = mNpcW_GET_WALK_NUM(common_data.save.now_npc_max);
+
+    for (i = 0; i < walk->info_max; i++) {
+        idx = mNpcW_DecideNpc(animals, walk->used_idx_bitfield);
+
+        mNpcW_SetNpcWalkInfo(info, &common_data.save.animals[idx], idx);
+        walk->used_idx_bitfield |= (1 << idx);
+        info++;
+    }
+
+    walk->used_info_num = walk->info_max;
+    mNpcW_InitGoalBlockSource(walk, animals);
+}
+
+s32 mNpcW_GetAppearStatusWay(u8* status, u8* way, Animal_c* animal) {
+    mNpc_walk_c* walk = &common_data.npc_walk;
+    s32 info_idx;
+    s32 res = FALSE;
+
+    *status = mNpcW_APPEAR_STATUS_NUM;
+
+    info_idx = mNpcW_GetNpcWalkInfoIdx(walk->info, mNpcW_MAX, &animal->id);
+
+    if (info_idx >= 0) {
+        *status = walk->info[info_idx].appear_info.status;
+        *way = walk->info[info_idx].appear_info.way;
+        res = TRUE;
+    } else {
+        *status = mNpcW_APPEAR_STATUS_0;
+        *way = mNpcW_APPEAR_WAY_NUM;
+    }
+
+    return res;
+}
+
+s32* mNpcW_GetArriveStayCountP(s32 idx) {
+    if (idx >= 0 && idx < ANIMAL_NUM_MAX) {
+        return &l_arrive_stay_count[idx];
+    }
+
+    return NULL;
+}
+
+s32 mNpcW_GetWalkInfoStatusGoalAnimalIdx(s32* status, s32* goal, s32 idx) {
+    s32 res = FALSE;
+
+    *status = mNpcW_INFO_STATUS_NUM;
+    *goal = mNpcW_GOAL_NUM;
+
+    if (idx >= 0 && idx < ANIMAL_NUM_MAX) {
+        mNpc_walk_c* walk = &common_data.npc_walk;
+        s32 info_idx = mNpcW_GetNpcWalkInfoIdx(walk->info, mNpcW_MAX, &common_data.save.animals[idx].id);
+
+        if (info_idx >= 0) {
+            res = TRUE;
+            *status = walk->info[info_idx].status;
+            *goal = walk->info[info_idx].goal_type;
+        }
+    }
+
+    return res;
+}
+
+s32 mNpcW_GetNearGate(s32* target_ut_x, s32* target_ut_z, s32 block_x, s32 block_z, s32 ut_x, s32 ut_z) {
+    u8 block_type;
+    mRF_gate_c* gate;
+    s32 difference;
+    UNUSED s32 pad;
+    s32 x;
+    s32 z;
+    s32 gate_count;
+    s32 i;
+    s32 res = FALSE;
+    s32 direction;
+
+    *target_ut_x = UT_X_NUM;
+    *target_ut_z = UT_Z_NUM;
```

Comment:
What does this even mean? What is a UT?

Reply:
Each acre is known as a block, in a block there's units that are referred as UT, so the specific x and z value are the units in the block

Reply:
And the units are different for x axis and z axis?

Reply:
Yeah x and z are different